Build failure: amdvlk

### Steps To Reproduce

Similar to Hydra now, I'm having issues with `amdvlk` with i686

### Build log

https://hydra.nixos.org/build/209322760

### Additional context

This is used for x86_64 to get the drivers to run Steam and other older games. Building from `master` as I have for months while I wait for #177623 + #187303 because I need to turn off the broken build flag. Maybe one day it will get reviewed and merged in and I can go back to unstable.

### Notify maintainers

@Flakebi 

### Metadata

Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 6.1.11, NixOS, 23.05 (Stoat), 23.05.20230213.dirty`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.13.2`
 - channels(toastal): `""`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
```

I experience the same problem:

```
ninja: build stopped: subcommand failed.
error: builder for '/nix/store/cb4pmz31hb8jd30l5qh9ny14cfwdg4bj-amdvlk-2022.Q4.4.drv' failed with exit code 1;
       last 10 log lines:
       > [196/2595] Building CXX object compiler/llpc/llvm/lib/Support/CMakeFiles/LLVMSupport.dir/Error.cpp.o
       > [197/2595] Building CXX object compiler/llpc/llvm/lib/Support/CMakeFiles/LLVMSupport.dir/FileOutputBuffer.cpp.o
       > [198/2595] Building CXX object compiler/llpc/llvm/lib/Support/CMakeFiles/LLVMSupport.dir/ELFAttributeParser.cpp.o
       > [199/2595] Building CXX object compiler/llpc/llvm/lib/Support/CMakeFiles/LLVMSupport.dir/FileUtilities.cpp.o
       > [200/2595] Building CXX object compiler/llpc/llvm/lib/Support/CMakeFiles/LLVMSupport.dir/FileCollector.cpp.o
       > [201/2595] Building CXX object compiler/llpc/llvm/lib/Support/CMakeFiles/LLVMSupport.dir/APInt.cpp.o
       > [202/2595] Building CXX object compiler/llpc/llvm/lib/Demangle/CMakeFiles/LLVMDemangle.dir/ItaniumDemangle.cpp.o
       > [203/2595] Building CXX object compiler/llpc/llvm/lib/Support/CMakeFiles/LLVMSupport.dir/APFloat.cpp.o
       > [204/2595] Building CXX object compiler/llpc/llvm/lib/Support/CMakeFiles/LLVMSupport.dir/CommandLine.cpp.o
       > ninja: build stopped: subcommand failed.
       For full logs, run 'nix log /nix/store/cb4pmz31hb8jd30l5qh9ny14cfwdg4bj-amdvlk-2022.Q4.4.drv'.
error: 1 dependencies of derivation '/nix/store/cjavca5yx86wmkkjljawifh62h82lacq-opengl-drivers-32bit.drv' failed to build
```

--- a/pkgs/top-level/all-packages.nix
+++ b/pkgs/top-level/all-packages.nix
@@ -24361,7 +24361,10 @@ with pkgs;
 
   directx-headers = callPackage ../development/libraries/directx-headers {};
 
-  directx-shader-compiler = callPackage ../tools/graphics/directx-shader-compiler {};
+  directx-shader-compiler = callPackage ../tools/graphics/directx-shader-compiler {
+    # https://github.com/NixOS/nixpkgs/issues/216294
+    stdenv = if stdenv.cc.isGNU && stdenv.isi686 then gcc11Stdenv else stdenv;
+  };
 
   dkimproxy = callPackage ../servers/mail/dkimproxy { };
 

